### PR TITLE
Fix getting invalid results for field access on optional record json field

### LINKED ISSUE #38840: [Bug]: Invalid result for field access expression on record json field
### Description

```
import ballerina/io;

type FooRec1 record {
    json j;
};

type FooRec3 record {
    json j?;
    FooRec1 fooRec?;
};

public function main() returns error? {
    FooRec3 rec2 = {j: "1", fooRec: {j: {k: "3", l: {m: "4"}}}};
    json val = check rec2.fooRec?.j.k;
    io:println(val); // got {"k":"3","l":{"m":"4"}} which is wrong
}
```
expected result for `val` is `3` but got  `{"k":"3","l":{"m":"4"}}` 

### Steps to Reproduce

_No response_

### Affected Version(s)

Master 
2201.3.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by ushirask**
Fixes #38840 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by prakanth97** [comment]
Shall we sync with the master. So, the codecov failure will pass

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -9848,7 +9848,9 @@ private void handleSafeNavigation(BLangBlockStmt blockStmt, BLangAccessExpressio
             if (this.safeNavigationAssignment != null) {
                 this.safeNavigationAssignment.expr = addConversionExprIfRequired(accessExpr, tempResultVar.getBType());
             }
-            return;
+            if (Types.getReferredType(originalType).tag != TypeTags.JSON) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:9851]: What's the plan for https://github.com/ballerina-platform/ballerina-lang/pull/39461/files#r1096973110?

Comment (ushirask) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:9851]: Can we use [isLaxFieldAccessAllowed](https://github.com/ballerina-platform/ballerina-lang/pull/39600/files#diff-51851f1451506fedd7c81b67a92282fada4eb00fe4728c70445d7262aed401f3R275) for this?

Comment (ushirask) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:9851]: @MaryamZi

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:9851]: We need to try and see. What I meant was that we need to revisit this entire block and do the correct thing properly.

**Comment by hasithaa** [comment]
Differed till the next RnD cycle (2024).

**Pull request closed without merging**